Automattic/jetpack
# RNA: export the Connection store

#### Changes proposed in this Pull Request:
- remove the `withConnectionStatus` HOC
- export the package store ID
- modify the Connection UI package to get rid of its own connection status object, and pull the required data from the RNA Connection package store
- modify Jetpack to put the connection status from its initial state into RNA Connection package
- remove connection status handling from Jetpack Backup's state and make it rely on RNA Connection

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
_The testing is long and tedious. No functional changes, but we need to make sure all the connection flows still work properly._

_If you're testing locally, make sure you rebuild `plugins/jetpack`, `plugins/backup`, and `packages/connection-ui`.

0. Read the corresponding section in the [README.md](https://github.com/Automattic/jetpack/tree/try/rna-connection-store-export/projects/js-packages/connection#fetching-connection-status-and-other-data-from-the-store). Does everything make sense and free of typos?

##### Jetpack
1. Go to Jetpack Dashboard, disconnect if connected. Confirm that you see the connection screen.
2. Open the "Network Monitor" and reload the connection screen. Confirm that the API request to `/wp-json/jetpack/v4/connection` does NOT get sent. The RNA Connection package receives the connection status from Jetpack's initial state, so that request is excessive.
3. Connect Jetpack, confirm everything's all right.
4. Disconnect Jetpack, connect in site-only mode, go to Jetpack Dashboard. Open the "Network Monitor", click "Connect your WP.com account", confirm the `/v4/connection` API request does NOT get send, and the connection screen shows up.
5. Connect the user, confirm it goes well.
6. Log in as a separate non-admin user, connect to WP.com, confirm it works as usual.

##### Jetpack Backup
1. Disconnect if connected. You should see the connection screen.
2. Open the "Network Monitor" and reload the connection screen. Confirm that the API request to `/wp-json/jetpack/v4/connection` DOES get sent. The backup plugin no longer stores connection status in its store, and RNA Connection package loads it via the API.
3. Connect to WP.com, confirm that it works fine.
4. After you're back in Jetpack Backup, confirm that the connection status card shows up correctly (you're connected, your name and avatar are there).
5. Disconnect, confirm that you're back to the connection screen.

##### Connection Manager (connection screen with a required plan)
_If testing using JN, you're going to need [this plugin](https://gist.github.com/sergeymitr/bc005e1b893e778c03e4f43debb91c30) installed and activated._

1. Go to "Tools -> Connection Manager". You should see the new Backup connection screen. Click on "Get Jetpack Backup".
2. Go through the connection flow, confirm it works well.
3. After you're back to the Connection Manager, confirm that the connection status card shows up correctly (you're connected, your name and avatar are there).
